adapter/internal/oasparser/model/policy_container.go

```diff
@@ -0,0 +1,251 @@
+			Message:   fmt.Sprintf("Operation policy validation failed, ignoring the policy \"%v\": %v", policy.PolicyName, err),
+			Severity:  logging.MINOR,
+			ErrorCode: 2204,
+		})
+		return policy, err
+	}
+
+	defRaw := p[policy.PolicyName].Definition.RawData
+	t, err := template.New("policy-def").Parse(string(defRaw))
+	if err != nil {
+		loggers.LoggerOasparser.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprintf("Error parsing the operation policy definition \"%v\" into go template: %v", policy.PolicyName, err),
+			Severity:  logging.MINOR,
+			ErrorCode: 2205,
+		})
+		return Policy{}, err
+	}
+
+	var out bytes.Buffer
+	err = t.Execute(&out, policy.Parameters)
+	if err != nil {
+		loggers.LoggerOasparser.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprintf("Error parsing operation policy definition \"%v\": %v", policy.PolicyName, err),
+			Severity:  logging.MINOR,
+			ErrorCode: 2206,
+		})
+		return Policy{}, err
+	}
+
+	def := PolicyDefinition{}
+	if err := yaml.Unmarshal(out.Bytes(), &def); err != nil {
+		loggers.LoggerOasparser.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprintf("Error parsing formalized operation policy definition \"%v\" into yaml: %v", policy.PolicyName, err),
+			Severity:  logging.MINOR,
+			ErrorCode: 2207,
+		})
+		return Policy{}, err
+	}
+
+	// update templated policy itself and return, not updating a pointer to keep the original template values as it is.
+	policy.Parameters = def.Definition.Parameters
+	policy.Action = def.Definition.Action
+
+	return policy, nil
+}
+
+// PolicySpecification holds policy specification from ./Policy/<policy>.yaml files
+type PolicySpecification struct {
+	Type    string `yaml:"type" json:"type"`
+	Version string `yaml:"version" json:"version"`
+	Data    struct {
+		Name              string   `yaml:"name"`
+		ApplicableFlows   []string `yaml:"applicableFlows"`
+		SupportedGateways []string `yaml:"supportedGateways"`
+		SupportedAPITypes []string `yaml:"supportedApiTypes"`
+		MultipleAllowed   bool     `yaml:"multipleAllowed"`
+		PolicyAttributes  []struct {
+			Name            string `yaml:"name"`
+			ValidationRegex string `yaml:"validationRegex,omitempty"`
+			Type            string `yaml:"type"`
+			Required        bool   `yaml:"required"`
+		} `yaml:"policyAttributes"`
+	}
+}
+
+// PolicyDefinition holds the content of policy definition which is rendered from ./Policy/<policy>.gotmpl files
+type PolicyDefinition struct {
+	Definition struct {
+		Action     string                 `yaml:"action"`
+		Parameters map[string]interface{} `yaml:"parameters"`
+	} `yaml:"definition"`
+	RawData []byte `yaml:"-"`
+}
+
+// validatePolicy validates the given policy against the spec
+func (spec *PolicySpecification) validatePolicy(policy Policy, flow PolicyFlow, stats map[string]policyStats, index int) error {
+	if spec.Data.Name != policy.PolicyName {
+		return fmt.Errorf("invalid policy specification, spec name %s and policy name %s mismatch", spec.Data.Name, policy.PolicyName)
```

Comment:
Can we improve these loggings? We need to identify the API, org, etc.  We can use the new logging methods.

Reply:
added with https://github.com/wso2/product-microgateway/pull/2656/commits/cb46ca154bdeb62c3ce0e3a3a24c0295f9aedcea